### PR TITLE
Ruby: refactor regex libraries

### PR DESCRIPTION
**Pull request by aibaars**
This pull request refactors the regular expression libraries for Ruby to match the Python and JavaScript libraries.

I aligned the contents of the regex parser module (`ruby/ql/lib/codeql/ruby/regexp/internal/ParseRegExp.qll`) and the regex AST module (`ruby/ql/lib/codeql/ruby/regexp/internal/RegExpTreeView.qll`) as much as possible with the respective Python implementations `python/ql/lib/semmle/python/regex.qll` and `python/ql/lib/semmle/python/RegexTreeView.qll`

The additional classes and predicates are moved into `ruby/ql/lib/codeql/ruby/Regexp.qll` .This module should be used by most users and  imports the parser module and re-exports the AST module.

### CODE REVIEW

**Review by Johnstone456 — COMMENTED** [review]
-

**Comment by Johnstone456** [comment]
> #8491

File: python/ql/lib/semmle/python/regex.qll [hunk 1]
```diff
@@ -314,8 +314,10 @@ abstract class RegexString extends Expr {
     result = this.(Bytes).getS()
   }
 
+  /** Gets the `i`th character of this regex */
   string getChar(int i) { result = this.getText().charAt(i) }
 
+  /** Gets the `i`th character of this regex, unless it is part of an character escape sequence. */
```
Comment (yoff) [inline python/ql/lib/semmle/python/regex.qll:320]:
```suggestion
  /** Gets the `i`th character of this regex, unless it is part of a character escape sequence. */
```

File: ruby/ql/lib/codeql/ruby/regexp/internal/ParseRegExp.qll [hunk 2]
```diff
@@ -229,8 +256,10 @@ abstract class RegExp extends AST::StringlikeLiteral {
   /** Gets the text of this regex */
   string getText() { result = this.getConstantValue().getString() }
 
+  /** Gets the `i`th character of this regex */
   string getChar(int i) { result = this.getText().charAt(i) }
 
+  /** Gets the `i`th character of this regex, unless it is part of an character escape sequence. */
```
Comment (yoff) [inline ruby/ql/lib/codeql/ruby/regexp/internal/ParseRegExp.qll:262]:
```suggestion
  /** Gets the `i`th character of this regex, unless it is part of a character escape sequence. */
```

**Comment by aibaars** [comment]
> LGTM, only typos. Why does `ReDoS.expected` change?

@yoff `ReDoS.expected` changed because I noticed we were missing handling of `\f` and `\v` when I compared the Python and the Ruby versions of the QLL files: 
https://github.com/github/codeql/pull/8489/files#diff-19c4ea6adc90edc0f6a1fa956d00494f6a23c1f42eab27f4c80ff5b195cdeccdR495-R497

File: ruby/ql/lib/codeql/ruby/Regexp.qll [hunk 3]
```diff
@@ -5,8 +5,8 @@
  * terms.
  */
 
-import regexp.RegExpTreeView // re-export
-private import regexp.ParseRegExp
+import regexp.internal.RegExpTreeView // re-export
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/Regexp.qll:8]: This appears dangerous: I would expect I could make all sorts of breaking changes to a file in an `internal` folder (except perhaps if it contains a `Public` module that documents it explicitly).

File: ruby/ql/lib/change-notes/2022-03-21-regex.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* The `ParseRegExp` and `RegExpTreeView` modules are now "internal" modules, users should use `codeql.ruby.Regexp` instead.
```
Comment (nickrolfe) [inline ruby/ql/lib/change-notes/2022-03-21-regex.md:4]: I know this is pedantic, but I dislike [comma splices](https://en.wikipedia.org/wiki/Comma_splice) 😛 
```suggestion
* The `ParseRegExp` and `RegExpTreeView` modules are now "internal" modules. Users should use `codeql.ruby.Regexp` instead.
```

File: ruby/ql/lib/codeql/ruby/Regexp.qll [hunk 5]
```diff
@@ -0,0 +1,143 @@
+/**
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/Regexp.qll:1]: Regarding the naming of this file: although the Ruby class is spelled `Regexp`, we've been using a capital E (`RegExp`) for the QL module/class names, for consistency between languages.

**Review by nickrolfe — COMMENTED** [review]
I have a couple of trivial/pedantic comments. Other than that, having two different `RegExpTreeView.qll` files seems undesirable. Can they be merged?

File: ruby/ql/lib/codeql/ruby/Regexp.qll (hunk 5 above)
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/Regexp.qll:1]: Yes, I considered that, however, for JavaScript the file is named `javascript/ql/lib/semmle/javascript/Regexp.qll` too.

**Comment by aibaars** [comment]
> Other than that, having two different `RegExpTreeView.qll` files seems undesirable. Can they be merged?

The one in the `performance` folder is required for the `ReDoS` queries. We should probably rename it to `ReDoSUtilSpecific.qll` or so.

**Review by yoff — APPROVED** [review]
LGTM

**Comment by yoff** [comment]
> Should we have done a DCA run before merging this PR? I am not intimate with the changes, but the diff is quite large.

I do not expect any of the changes to have an impact, but yes..

**Pull request merged**